File: Go/isbn_utils/isbn_utils_test.go
package isbn_utils

import (
	"testing"
)

func TestClean(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"978-3-16-148410-0", "9783161484100"},
		{"0-306-40615-2", "0306406152"},
		{"ISBN 0-306-40615-2", "0306406152"},
		{"978 3 16 148410 0", "9783161484100"},
		{"0306406152", "0306406152"},
		{"0-19-852663-6", "0198526636"},
		{"0-8044-2957-X", "080442957X"},
	}

	for _, tt := range tests {
		result := Clean(tt.input)
		if result != tt.expected {
			t.Errorf("Clean(%q) = %q, want %q", tt.input, result, tt.expected)
		}
	}
}

func TestValidateISBN10(t *testing.T) {
	tests := []struct {
		isbn     string
		expected bool
	}{
		// Valid ISBN-10s
		{"0306406152", true},      // Valid
		{"0-306-40615-2", true},   // Valid with hyphens
		{"080442957X", true},      // Valid with X check digit
		{"0-8044-2957-X", true},   // Valid with X and hyphens
		{"0198526636", true},      // Valid
		{"0596007973", true},      // Valid
		{"0321334876", true},      // Valid
		
		// Invalid ISBN-10s
		{"0306406153", false},     // Wrong check digit
		{"0804429570", false},     // Wrong check digit (should be X)
		{"123456789", false},      // Too short
		{"12345678901", false},    // Too long
		{"ABCDEFGHIJ", false},     // Non-numeric
		{"", false},               // Empty
	}

	for _, tt := range tests {
		valid, _ := ValidateISBN10(tt.isbn)
		if valid != tt.expected {
			t.Errorf("ValidateISBN10(%q) = %v, want %v", tt.isbn, valid, tt.expected)
		}
	}
}

func TestValidateISBN13(t *testing.T) {
	tests := []struct {
		isbn     string
		expected bool
	}{
		// Valid ISBN-13s
		{"9783161484100", true},      // Valid
		{"978-3-16-148410-0", true},  // Valid with hyphens
		{"9780306406157", true},      // Valid
		{"9780596007977", true},      // Valid
		{"9791091146135", true},      // Valid 979 prefix
		{"9780321334876", true},      // Valid
		
		// Invalid ISBN-13s
		{"9783161484101", false},     // Wrong check digit
		{"9780306406158", false},     // Wrong check digit
		{"123456789012", false},      // Too short
		{"12345678901234", false},   // Too long
		{"ABCDEFGHIJKLM", false},    // Non-numeric
		{"", false},                  // Empty
	}

	for _, tt := range tests {
		valid, _ := ValidateISBN13(tt.isbn)
		if valid != tt.expected {
			t.Errorf("ValidateISBN13(%q) = %v, want %v", tt.isbn, valid, tt.expected)
		}
	}
}

func TestValidate(t *testing.T) {
	tests := []struct {
		isbn     string
		expected bool
	}{
		{"0306406152", true},        // Valid ISBN-10
		{"9783161484100", true},     // Valid ISBN-13
		{"0-306-40615-2", true},     // Valid ISBN-10 with hyphens
		{"080442957X", true},        // Valid ISBN-10 with X
		{"12345", false},            // Invalid length
		{"", false},                 // Empty
	}

	for _, tt := range tests {
		valid, _ := Validate(tt.isbn)
		if valid != tt.expected {
			t.Errorf("Validate(%q) = %v, want %v", tt.isbn, valid, tt.expected)
		}
	}
}

func TestGenerateCheckDigit10(t *testing.T) {
	tests := []struct {
		prefix   string
		expected string
		hasError bool
	}{
		{"030640615", "2", false},
		{"080442957", "X", false},
		{"019852663", "6", false},
		{"059600797", "3", false},
		{"12345678", "", true},   // Too short
		{"1234567890", "", true}, // Too long
		{"ABCDEFGHI", "", true},  // Non-numeric
	}

	for _, tt := range tests {
		result, err := GenerateCheckDigit10(tt.prefix)
		if tt.hasError {
			if err == nil {
				t.Errorf("GenerateCheckDigit10(%q) expected error, got none", tt.prefix)
			}
		} else {
			if err != nil {
				t.Errorf("GenerateCheckDigit10(%q) unexpected error: %v", tt.prefix, err)
			}
			if result != tt.expected {
				t.Errorf("GenerateCheckDigit10(%q) = %q, want %q", tt.prefix, result, tt.expected)
			}
		}
	}
}

func TestGenerateCheckDigit13(t *testing.T) {
	tests := []struct {
		prefix   string
		expected string
		hasError bool
	}{
		{"978316148410", "0", false},
		{"978030640615", "7", false},
		{"978059600797", "7", false},
		{"978032133487", "6", false},
		{"12345678901", "", true},   // Too short
		{"1234567890123", "", true}, // Too long
		{"ABCDEFGHIJKL", "", true},  // Non-numeric
	}

	for _, tt := range tests {
		result, err := GenerateCheckDigit13(tt.prefix)
		if tt.hasError {
			if err == nil {
				t.Errorf("GenerateCheckDigit13(%q) expected error, got none", tt.prefix)
			}
		} else {
			if err != nil {
				t.Errorf("GenerateCheckDigit13(%q) unexpected error: %v", tt.prefix, err)
			}
			if result != tt.expected {
				t.Errorf("GenerateCheckDigit13(%q) = %q, want %q", tt.prefix, result, tt.expected)
			}
		}
	}
}

func TestToISBN13(t *testing.T) {
	tests := []struct {
		isbn10   string
		expected string
		hasError bool
	}{
		{"0306406152", "9780306406157", false},
		{"0-306-40615-2", "9780306406157", false},
		{"080442957X", "9780804429571", false},
		{"0-8044-2957-X", "9780804429571", false},
		{"0198526636", "9780198526637", false},
		{"12345", "", true},  // Invalid length
	}

	for _, tt := range tests {
		result, err := ToISBN13(tt.isbn10)
		if tt.hasError {
			if err == nil {
				t.Errorf("ToISBN13(%q) expected error, got none", tt.isbn10)
			}
		} else {
			if err != nil {
				t.Errorf("ToISBN13(%q) unexpected error: %v", tt.isbn10, err)
			}
			if result != tt.expected {
				t.Errorf("ToISBN13(%q) = %q, want %q", tt.isbn10, result, tt.expected)
			}
		}
	}
}

func TestToISBN10(t *testing.T) {
	tests := []struct {
		isbn13   string
		expected string
		hasError bool
	}{
		{"9780306406157", "0306406152", false},
		{"978-0-306-40615-7", "0306406152", false},
		{"9780804429571", "080442957X", false},
		{"9780198526637", "0198526636", false},
		{"9791091146135", "", true},  // 979 prefix cannot be converted to ISBN-10
		{"12345", "", true},           // Invalid length
	}

	for _, tt := range tests {
		result, err := ToISBN10(tt.isbn13)
		if tt.hasError {
			if err == nil {
				t.Errorf("ToISBN10(%q) expected error, got none", tt.isbn13)
			}
		} else {
			if err != nil {
				t.Errorf("ToISBN10(%q) unexpected error: %v", tt.isbn13, err)
			}
			if result != tt.expected {
				t.Errorf("ToISBN10(%q) = %q, want %q", tt.isbn13, result, tt.expected)
			}
		}
	}
}

func TestFormat(t *testing.T) {
	tests := []struct {
		isbn     string
		expected string
	}{
		{"0306406152", "0-30640-615-2"},
		{"9783161484100", "978-3-16148-410-0"},
		{"080442957X", "0-80442-957-X"},
		{"9780306406157", "978-0-30640-615-7"},
	}

	for _, tt := range tests {
		result := Format(tt.isbn)
		if result != tt.expected {
			t.Errorf("Format(%q) = %q, want %q", tt.isbn, result, tt.expected)
		}
	}
}

func TestGetType(t *testing.T) {
	tests := []struct {
		isbn         string
		expectedType string
		hasError     bool
	}{
		{"0306406152", TypeISBN10, false},
		{"9783161484100", TypeISBN13, false},
		{"12345", "", true},
	}

	for _, tt := range tests {
		result, err := GetType(tt.isbn)
		if tt.hasError {
			if err == nil {
				t.Errorf("GetType(%q) expected error, got none", tt.isbn)
			}
		} else {
			if err != nil {
				t.Errorf("GetType(%q) unexpected error: %v", tt.isbn, err)
			}
			if result != tt.expectedType {
				t.Errorf("GetType(%q) = %q, want %q", tt.isbn, result, tt.expectedType)
			}
		}
	}
}

func TestIsISBN(t *testing.T) {
	if !IsISBN("0306406152") {
		t.Error("IsISBN('0306406152') should be true")
	}
	if !IsISBN("9783161484100") {
		t.Error("IsISBN('9783161484100') should be true")
	}
	if IsISBN("12345") {
		t.Error("IsISBN('12345') should be false")
	}
}

func TestIsISBN10(t *testing.T) {
	if !IsISBN10("0306406152") {
		t.Error("IsISBN10('0306406152') should be true")
	}
	if IsISBN10("9783161484100") {
		t.Error("IsISBN10('9783161484100') should be false")
	}
}

func TestIsISBN13(t *testing.T) {
	if !IsISBN13("9783161484100") {
		t.Error("IsISBN13('9783161484100') should be true")
	}
	if IsISBN13("0306406152") {
		t.Error("IsISBN13('0306406152') should be false")
	}
}

func TestNormalize(t *testing.T) {
	tests := []struct {
		isbn     string
		expected string
		hasError bool
	}{
		{"0306406152", "9780306406157", false},    // ISBN-10 to ISBN-13
		{"9783161484100", "9783161484100", false}, // Already ISBN-13
		{"12345", "", true},                       // Invalid
	}

	for _, tt := range tests {
		result, err := Normalize(tt.isbn)
		if tt.hasError {
			if err == nil {
				t.Errorf("Normalize(%q) expected error, got none", tt.isbn)
			}
		} else {
			if err != nil {
				t.Errorf("Normalize(%q) unexpected error: %v", tt.isbn, err)
			}
			if result != tt.expected {
				t.Errorf("Normalize(%q) = %q, want %q", tt.isbn, result, tt.expected)
			}
		}
	}
}

func TestGenerateISBN13(t *testing.T) {
	result, err := GenerateISBN13("978316148410")
	if err != nil {
		t.Errorf("GenerateISBN13 unexpected error: %v", err)
	}
	if result != "9783161484100" {
		t.Errorf("GenerateISBN13 = %q, want %q", result, "9783161484100")
	}

	_, err = GenerateISBN13("12345")
	if err == nil {
		t.Error("GenerateISBN13 should return error for invalid prefix length")
	}
}

func TestGenerateISBN10(t *testing.T) {
	result, err := GenerateISBN10("030640615")
	if err != nil {
		t.Errorf("GenerateISBN10 unexpected error: %v", err)
	}
	if result != "0306406152" {
		t.Errorf("GenerateISBN10 = %q, want %q", result, "0306406152")
	}

	result, err = GenerateISBN10("080442957")
	if err != nil {
		t.Errorf("GenerateISBN10 unexpected error: %v", err)
	}
	if result != "080442957X" {
		t.Errorf("GenerateISBN10 = %q, want %q", result, "080442957X")
	}

	_, err = GenerateISBN10("12345678")
	if err == nil {
		t.Error("GenerateISBN10 should return error for invalid prefix length")
	}
}

func TestParse(t *testing.T) {
	// Test ISBN-10
	isbn10, err := Parse("0306406152")
	if err != nil {
		t.Errorf("Parse ISBN-10 unexpected error: %v", err)
	}
	if isbn10.Type != TypeISBN10 {
		t.Errorf("Parse ISBN-10 type = %q, want %q", isbn10.Type, TypeISBN10)
	}
	if isbn10.Number != "0306406152" {
		t.Errorf("Parse ISBN-10 number = %q, want %q", isbn10.Number, "0306406152")
	}
	if isbn10.Check != "2" {
		t.Errorf("Parse ISBN-10 check = %q, want %q", isbn10.Check, "2")
	}

	// Test ISBN-13
	isbn13, err := Parse("9783161484100")
	if err != nil {
		t.Errorf("Parse ISBN-13 unexpected error: %v", err)
	}
	if isbn13.Type != TypeISBN13 {
		t.Errorf("Parse ISBN-13 type = %q, want %q", isbn13.Type, TypeISBN13)
	}
	if isbn13.Number != "9783161484100" {
		t.Errorf("Parse ISBN-13 number = %q, want %q", isbn13.Number, "9783161484100")
	}
	if isbn13.Prefix != "978" {
		t.Errorf("Parse ISBN-13 prefix = %q, want %q", isbn13.Prefix, "978")
	}

	// Test invalid
	_, err = Parse("12345")
	if err == nil {
		t.Error("Parse should return error for invalid ISBN")
	}
}

// Benchmark tests
func BenchmarkValidateISBN10(b *testing.B) {
	for i := 0; i < b.N; i++ {
		ValidateISBN10("0306406152")
	}
}

func BenchmarkValidateISBN13(b *testing.B) {
	for i := 0; i < b.N; i++ {
		ValidateISBN13("9783161484100")
	}
}

func BenchmarkToISBN13(b *testing.B) {
	for i := 0; i < b.N; i++ {
		ToISBN13("0306406152")
	}
}

func BenchmarkFormat(b *testing.B) {
	for i := 0; i < b.N; i++ {
		Format("9783161484100")
	}
}

File: Python/isbn_utils/isbn_utils_test.py
#!/usr/bin/env python3
"""
ISBN Utilities Test Suite

Tests for ISBN-10 and ISBN-13 validation, parsing, formatting, and conversion.
"""

import sys
import os

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from isbn_utils.mod import (
    clean, validate, validate_isbn10, validate_isbn13,
    calculate_check_digit_isbn10, calculate_check_digit_isbn13,
    isbn10_to_isbn13, isbn13_to_isbn10,
    parse, format, get_type, is_isbn10, is_isbn13,
    generate_isbn10, generate_isbn13, batch_validate,
    normalize, find_isbns, compare, get_group_name, analyze, repair
)


def test_clean():
    """Test ISBN cleaning function."""
    print("Testing clean()...")
    
    # Basic cleaning
    assert clean('978-0-306-40615-7') == '9780306406157'
    assert clean('0 306 40615 2') == '0306406152'
    assert clean('9780306406157') == '9780306406157'
    
    # With X check digit
    assert clean('0-8044-2957-X') == '080442957X'
    assert clean('0-8044-2957-x') == '080442957X'  # Uppercase conversion
    
    print("  ✓ clean() tests passed")


def test_validate_isbn10():
    """Test ISBN-10 validation."""
    print("Testing validate_isbn10()...")
    
    # Valid ISBN-10s
    assert validate_isbn10('0306406152') == True
    assert validate_isbn10('0-306-40615-2') == True
    assert validate_isbn10('080442957X') == True  # X check digit
    assert validate_isbn10('0-8044-2957-X') == True
    assert validate_isbn10('0471958697') == True
    assert validate_isbn10('0-471-95869-7') == True
    assert validate_isbn10('0851310419') == True
    
    # Invalid ISBN-10s
    assert validate_isbn10('0306406151') == False  # Wrong check digit
    assert validate_isbn10('030640615') == False   # Too short
    assert validate_isbn10('03064061523') == False # Too long
    assert validate_isbn10('ABCDEFGHIJ') == False  # Non-numeric
    
    print("  ✓ validate_isbn10() tests passed")


def test_validate_isbn13():
    """Test ISBN-13 validation."""
    print("Testing validate_isbn13()...")
    
    # Valid ISBN-13s
    assert validate_isbn13('9780306406157') == True
    assert validate_isbn13('978-0-306-40615-7') == True
    assert validate_isbn13('9790306406156') == True
    assert validate_isbn13('9780471958697') == True
    assert validate_isbn13('9780140283297') == True
    
    # Invalid ISBN-13s
    assert validate_isbn13('9780306406158') == False  # Wrong check digit
    assert validate_isbn13('978030640615') == False   # Too short
    assert validate_isbn13('97803064061577') == False # Too long
    
    print("  ✓ validate_isbn13() tests passed")


def test_validate():
    """Test general validate function."""
    print("Testing validate()...")
    
    # Valid ISBN-10
    assert validate('0306406152') == True
    assert validate('0-306-40615-2') == True
    
    # Valid ISBN-13
    assert validate('9780306406157') == True
    assert validate('978-0-306-40615-7') == True
    
    # Invalid
    assert validate('invalid') == False
    assert validate('123456789') == False
    
    print("  ✓ validate() tests passed")


def test_check_digit_calculation():
    """Test check digit calculation."""
    print("Testing check digit calculation...")
    
    # ISBN-10
    assert calculate_check_digit_isbn10('030640615') == '2'
    assert calculate_check_digit_isbn10('080442957') == 'X'  # X check digit
    assert calculate_check_digit_isbn10('047195869') == '7'
    assert calculate_check_digit_isbn10('047111709') == '9'
    
    # ISBN-13
    assert calculate_check_digit_isbn13('978030640615') == '7'
    assert calculate_check_digit_isbn13('979030640615') == '6'
    assert calculate_check_digit_isbn13('978047195869') == '7'
    
    print("  ✓ Check digit calculation tests passed")


def test_isbn10_to_isbn13():
    """Test ISBN-10 to ISBN-13 conversion."""
    print("Testing isbn10_to_isbn13()...")
    
    # Standard conversions
    assert isbn10_to_isbn13('0306406152') == '9780306406157'
    assert isbn10_to_isbn13('0-306-40615-2') == '9780306406157'
    assert isbn10_to_isbn13('0471958697') == '9780471958697'
    assert isbn10_to_isbn13('080442957X') == '9780804429573'
    
    # Validate converted ISBN-13s
    assert validate_isbn13(isbn10_to_isbn13('0306406152')) == True
    
    print("  ✓ isbn10_to_isbn13() tests passed")


def test_isbn13_to_isbn10():
    """Test ISBN-13 to ISBN-10 conversion."""
    print("Testing isbn13_to_isbn10()...")
    
    # Standard conversions (978 prefix only)
    assert isbn13_to_isbn10('9780306406157') == '0306406152'
    assert isbn13_to_isbn10('978-0-306-40615-7') == '0306406152'
    assert isbn13_to_isbn10('9780471958697') == '0471958697'
    assert isbn13_to_isbn10('9780804429573') == '080442957X'
    
    # 979 prefix cannot be converted
    assert isbn13_to_isbn10('9790306406156') == None
    
    # Validate converted ISBN-10s
    assert validate_isbn10(isbn13_to_isbn10('9780306406157')) == True
    
    print("  ✓ isbn13_to_isbn10() tests passed")


def test_parse():
    """Test ISBN parsing."""
    print("Testing parse()...")
    
    # Parse ISBN-10
    result = parse('0306406152')
    assert result['valid'] == True
    assert result['type'] == 'ISBN-10'
    assert result['check_digit'] == '2'
    assert result['isbn10'] == '0306406152'
    assert result['isbn13'] == '9780306406157'
    
    # Parse ISBN-13
    result = parse('9780306406157')
    assert result['valid'] == True
    assert result['type'] == 'ISBN-13'
    assert result['prefix'] == '978'
    assert result['check_digit'] == '7'
    assert result['isbn13'] == '9780306406157'
    assert result['isbn10'] == '0306406152'
    
    # Parse ISBN-13 with 979 prefix
    result = parse('9790306406156')
    assert result['valid'] == True
    assert result['prefix'] == '979'
    assert result['isbn10'] == None  # Cannot convert to ISBN-10
    
    # Parse invalid
    result = parse('invalid')
    assert result['valid'] == False
    assert result['type'] == None
    
    print("  ✓ parse() tests passed")


def test_format():
    """Test ISBN formatting."""
    print("Testing format()...")
    
    # ISBN-10 formatting
    assert format('0306406152', 'clean') == '0306406152'
    assert format('0306406152', 'hyphenated') == '0-30640-615-2'
    assert format('0306406152', 'spaced') == '0 30640 615 2'
    
    # ISBN-13 formatting
    assert format('9780306406157', 'clean') == '9780306406157'
    assert format('9780306406157', 'hyphenated') == '978-0-3064-0615-7'
    assert format('9780306406157', 'spaced') == '978 0 3064 0615 7'
    
    print("  ✓ format() tests passed")


def test_get_type():
    """Test get_type function."""
    print("Testing get_type()...")
    
    assert get_type('0306406152') == 'ISBN-10'
    assert get_type('9780306406157') == 'ISBN-13'
    assert get_type('invalid') == None
    
    print("  ✓ get_type() tests passed")


def test_is_isbn10_is_isbn13():
    """Test is_isbn10 and is_isbn13 functions."""
    print("Testing is_isbn10() and is_isbn13()...")
    
    # is_isbn10
    assert is_isbn10('0306406152') == True
    assert is_isbn10('9780306406157') == False
    
    # is_isbn13
    assert is_isbn13('9780306406157') == True
    assert is_isbn13('0306406152') == False
    
    print("  ✓ is_isbn10() and is_isbn13() tests passed")


def test_generate():
    """Test ISBN generation."""
    print("Testing generate_isbn10() and generate_isbn13()...")
    
    # Generate ISBN-10
    isbn10 = generate_isbn10()
    assert len(isbn10) == 10
    assert validate_isbn10(isbn10) == True
    
    # Generate ISBN-13
    isbn13 = generate_isbn13()
    assert len(isbn13) == 13
    assert validate_isbn13(isbn13) == True
    
    # Generate with prefix
    isbn13_978 = generate_isbn13(prefix='978')
    assert isbn13_978.startswith('978')
    assert validate_isbn13(isbn13_978) == True
    
    isbn13_979 = generate_isbn13(prefix='979')
    assert isbn13_979.startswith('979')
    assert validate_isbn13(isbn13_979) == True
    
    # Generate multiple and verify uniqueness
    isbn10s = [generate_isbn10() for _ in range(100)]
    assert len(set(isbn10s)) > 90  # Most should be unique
    
    print("  ✓ ISBN generation tests passed")


def test_batch_validate():
    """Test batch validation."""
    print("Testing batch_validate()...")
    
    isbns = [
        '0306406152',      # Valid ISBN-10
        '9780306406157',   # Valid ISBN-13
        'invalid',         # Invalid
        '080442957X',      # Valid ISBN-10 with X
        '9790306406156'    # Valid ISBN-13 with 979
    ]
    
    result = batch_validate(isbns)
    
    assert result['total'] == 5
    assert result['valid_count'] == 4
    assert result['invalid_count'] == 1
    assert result['isbn10_count'] == 2
    assert result['isbn13_count'] == 2
    
    print("  ✓ batch_validate() tests passed")


def test_normalize():
    """Test normalize function."""
    print("Testing normalize()...")
    
    # ISBN-10 to ISBN-13
    assert normalize('0306406152', 'isbn13') == '9780306406157'
    
    # ISBN-13 to ISBN-10
    assert normalize('9780306406157', 'isbn10') == '0306406152'
    
    # To clean
    assert normalize('978-0-306-40615-7', 'clean') == '9780306406157'
    
    # 979 prefix cannot convert to ISBN-10
    assert normalize('9790306406156', 'isbn10') == None
    
    print("  ✓ normalize() tests passed")


def test_find_isbns():
    """Test finding ISBNs in text."""
    print("Testing find_isbns()...")
    
    text = "Book ISBN: 978-0-306-40615-7 and also 0-306-40615-2"
    isbns = find_isbns(text)
    
    assert len(isbns) >= 2  # Should find both
    
    print("  ✓ find_isbns() tests passed")


def test_compare():
    """Test ISBN comparison."""
    print("Testing compare()...")
    
    # Same book, different formats
    result = compare('0306406152', '9780306406157')
    assert result['isbn1_valid'] == True
    assert result['isbn2_valid'] == True
    assert result['same_book'] == True
    assert result['same_isbn'] == False
    
    # Different books
    result = compare('0306406152', '0471958697')
    assert result['same_book'] == False
    assert result['same_isbn'] == False
    
    # Same ISBN
    result = compare('0306406152', '0-306-40615-2')
    assert result['same_isbn'] == True
    
    print("  ✓ compare() tests passed")


def test_get_group_name():
    """Test group name lookup."""
    print("Testing get_group_name()...")
    
    assert get_group_name('0') == 'English (UK/US)'
    assert get_group_name('1') == 'English (UK/US)'
    assert get_group_name('7') == 'China'
    assert get_group_name('4') == 'Japanese'
    assert get_group_name('999') == None  # Unknown
    
    print("  ✓ get_group_name() tests passed")


def test_analyze():
    """Test comprehensive analysis."""
    print("Testing analyze()...")
    
    result = analyze('978-0-306-40615-7')
    
    assert result['valid'] == True
    assert result['type'] == 'ISBN-13'
    assert result['cleaned'] == '9780306406157'
    assert result['isbn10'] == '0306406152'
    assert result['isbn13'] == '9780306406157'
    assert result['group'] == '0'
    assert result['group_name'] == 'English (UK/US)'
    assert 'clean' in result['formats']
    assert 'hyphenated' in result['formats']
    
    print("  ✓ analyze() tests passed")


def test_repair():
    """Test ISBN repair."""
    print("Testing repair()...")
    
    # Already valid
    assert repair('0306406152') == '0306406152'
    
    # Missing check digit (ISBN-10)
    assert repair('030640615') == '0306406152'
    
    # Missing check digit (ISBN-13)
    assert repair('978030640615') == '9780306406157'
    
    # Extra digit
    assert repair('03064061523') == '0306406152'  # 11 digits
    
    print("  ✓ repair() tests passed")


def test_real_world_isbns():
    """Test with real-world ISBN examples."""
    print("Testing real-world ISBNs...")
    
    # Famous books with known ISBNs
    real_isbns = [
        ('9780140283297', 'ISBN-13'),  # To Kill a Mockingbird
        ('9780061120084', 'ISBN-13'),  # To Kill a Mockingbird (another edition)
        ('9780451524935', 'ISBN-13'),  # 1984
        ('9780743273565', 'ISBN-13'),  # The Great Gatsby
    ]
    
    for isbn, expected_type in real_isbns:
        cleaned = clean(isbn)
        assert validate(cleaned) == True, f"Failed to validate {isbn}"
        assert get_type(cleaned) == expected_type, f"Wrong type for {isbn}"
    
    print("  ✓ Real-world ISBN tests passed")


def test_x_check_digit():
    """Test ISBN-10 with X check digit."""
    print("Testing X check digit...")
    
    # ISBN-10s that end with X
    x_isbns = [
        '080442957X',   # Valid ISBN-10 with X
        '0306406152',   # Valid ISBN-10 with digit check
    ]
    
    for isbn in x_isbns:
        cleaned = clean(isbn)
        assert validate_isbn10(cleaned) == True, f"Failed to validate {isbn}"
        
        # Convert to ISBN-13 and back (if 978 prefix compatible)
        isbn13 = isbn10_to_isbn13(cleaned)
        assert validate_isbn13(isbn13) == True
        
        # Convert back
        isbn10_back = isbn13_to_isbn10(isbn13)
        assert isbn10_back == cleaned
    
    print("  ✓ X check digit tests passed")


def run_all_tests():
    """Run all test functions."""
    print("=" * 60)
    print("ISBN Utilities Test Suite")
    print("=" * 60)
    print()
    
    tests = [
        test_clean,
        test_validate_isbn10,
        test_validate_isbn13,
        test_validate,
        test_check_digit_calculation,
        test_isbn10_to_isbn13,
        test_isbn13_to_isbn10,
        test_parse,
        test_format,
        test_get_type,
        test_is_isbn10_is_isbn13,
        test_generate,
        test_batch_validate,
        test_normalize,
        test_find_isbns,
        test_compare,
        test_get_group_name,
        test_analyze,
        test_repair,
        test_real_world_isbns,
        test_x_check_digit,
    ]
    
    for test in tests:
        try:
            test()
        except AssertionError as e:
            print(f"  ✗ Test failed: {test.__name__}")
            print(f"    Error: {e}")
            return False
        except Exception as e:
            print(f"  ✗ Test error: {test.__name__}")
            print(f"    Error: {e}")
            return False
    
    print()
    print("=" * 60)
    print("All tests passed! ✓")
    print("=" * 60)
    return True


if __name__ == '__main__':
    success = run_all_tests()
    sys.exit(0 if success else 1)